### ISSUE TITLE
Unable to install on 4k drive

### ISSUE THREAD

**Issue by bcg62** [coreos/fedora-coreos-tracker#385]
# Bug #

## Host Operating System Version ##
31.20200127.3.0
## Target Operating System Version ##
31.20200127.3.0
## coreos-installer Version ##
coreos-installer 0.1.2
## Expected Behavior ##
boot drive part partitioned
## Actual Behavior ##
```
Clearing partition table
Error: install failed
```
## Reproduction Steps ##

```
coreos-installer install /dev/sda --ignition /tmp/coreos-installer-j3u2Cx --firstboot-args rd.neednet=1 ip=dhcp  --image-url http://cache/assets/fedora-coreos/fedora-coreos-31.20200127.3.0-metal.x86_64.raw.xz
```

Is there a way to get better debug information to find out why or how this is failing?
```
coreos-installer install /dev/sda --ignition /tmp/coreos-installer-j3u2Cx --firstboot-args rd.neednet=1 ip=dhcp  --image-url http://cache/assets/fedora-coreos/fedora-coreos-31.20200127.3.0-metal.x86_64.raw.xz
Downloading image from http://cache/assets/fedora-coreos/fedora-coreos-31.20200127.3.0-metal.x86_64.raw.xz
Downloading signature from http://cache/assets/fedora-coreos/fedora-coreos-31.20200127.3.0-metal.x86_64.raw.xz.sig
gpg: Signature made Mon Feb 10 23:11:29 2020 UTC
gpg:                using RSA key 50CB390B3C3359C4
gpg: Good signature from "Fedora (31) <fedora-31-primary@fedoraproject.org>" [ultimate]
> Read disk 434.8 MiB/434.8 MiB (100%)
Error: couldn't find boot device for /dev/sda
Clearing partition table
Error: install failed
```

**Comment by bcg62** [comment]
reading though #18 is seems like there isn't currently a work around?

Is there a way or documentation on how someone might be able to build their own 4k compat images?

**Comment by dustymabe** [comment]
> Ahh I think that might be the issue. My understanding is that 4k disks will require separate metal images which we do not yet produce. Our current metal images assume traditional 512b sector-sized drives.

I almost wonder if we could handle this case as part of the "complex root devices" work. i.e. rather than make a separate metal image we act like we would if someone requested a different filesystem on ignition during boot. There's probably a million reasons why this wouldn't work, but I'd really like to not have to ship a separate set of metal images :(

**Comment by jlebon** [comment]
I started on this in https://github.com/coreos/coreos-assembler/pull/1130. If you're feeling adventurous and want to test that patch, you can build cosa with it, and then build FCOS as described in the [cosa README](https://github.com/coreos/coreos-assembler). :)

> I almost wonder if we could handle this case as part of the "complex root devices" work. i.e. rather than make a separate metal image we act like we would if someone requested a different filesystem on ignition during boot.

For 4k native setups, I don't think this is doable on first boot, because the system wouldn't even be able to boot in the first place. But it might be doable to do at install time via coreos-installer. E.g. it could detect that the target disk is 4k and translate the metal image on the fly? Not sure how feasible that is though. My understanding is that it's not just a bit-level transformation, but would require changing the filesystem level itself since it assumes 512b sectors (see https://github.com/coreos/fedora-coreos-tracker/issues/18#issuecomment-428321864). Worth exploring the idea though with some storage folks.

**Comment by darkmuggle** [comment]
> For 4k native setups, I don't think this is doable on first boot, because the system wouldn't even be able to boot in the first place. But it might be doable to do at install time via coreos-installer. E.g. it could detect that the target disk is 4k and translate the metal image on the fly? Not sure how feasible that is though. My understanding is that it's not just a bit-level transformation, but would require changing the filesystem level itself since it assumes 512b sectors (see [#18 (comment)](https://github.com/coreos/fedora-coreos-tracker/issues/18#issuecomment-428321864)). Worth exploring the idea though with some storage folk

I agree with @jlebon here. The installer will splat the disk down and then the complicated disk setup work won't even come to play and worse, prepping /boot is out of scope for complicated disks. For the time being, we'll need a new disk target.

**Comment by bgilbert** [comment]
https://github.com/coreos/coreos-installer/pull/167 teaches coreos-installer to detect this case and fail with a clearer error message.

**Comment by bgilbert** [comment]
Just did a quick experiment with `losetup --sector-size 4096` and ext4, XFS, and FAT32.  XFS filesystems created for a 512-byte sector fail to mount from a 4k-sector disk:

```
XFS (loop0p1): device supports 4096 byte sectors (not 512)
```

FAT32 likewise fails:

```
FAT-fs (loop0p3): logical sector size too small for device (logical sector size = 512)
```

ext4 works fine, and all three are fine if created on a 4k-sector disk but mounted from a 512-byte sector disk.

That's not an argument that we should just use 4k sectors for everything, though.  XFS, at least, is presumably relying on atomic sector writes.

**Comment by jlebon** [comment]
Yeah, I shouldn't have closed this as part of the cosa commit. There are a few things left to do before it's all plumbed through:
- plumb `metal4k` through all the metadata and website
- adjust coreos-installer to automatically use the `metal4k` image

**Comment by jlebon** [comment]
To start with:

https://github.com/coreos/fedora-coreos-pipeline/pull/210
https://github.com/coreos/fedora-coreos-browser/pull/6

For the stream metadata, how should we add it? We have `metal` right now, which encompasses the metal image (`raw.xz`), PXE artifacts (`pxe.{kernel,initramfs}`), and the live ISO (`iso`). I think it should be under `metal` still since the PXE and ISO artifacts don't care about sector sizes, but the `raw.xz` makes this awkward. Let's do `4k.raw.xz`, I guess?

**Comment by jlebon** [comment]
Hmm, or maybe a cleaner alternative is having a `metal.4k` object, which inside has `"raw.xz"`. OK, did that in https://github.com/coreos/fedora-coreos-releng-automation/pull/87.

**Comment by lucab** [comment]
@jlebon from the point of `coreos-installer list-stream` (the final consumer of all of this), how do the 4k artifacts fit in the arch-platform-format tuple?

**Comment by jlebon** [comment]
@lucab In https://github.com/coreos/fedora-coreos-releng-automation/pull/87, I went back to just using `4k.raw.xz`. So the `list-stream` output would look like this:

```
Architecture  Platform   Format
x86_64        aliyun     qcow2.xz
x86_64        aws        vmdk.xz
x86_64        azure      vhd.xz
x86_64        exoscale   qcow2.xz
x86_64        gcp        tar.gz
x86_64        metal      iso
x86_64        metal      pxe
x86_64        metal      raw.xz
x86_64        metal      4k.raw.xz
x86_64        openstack  qcow2.xz
x86_64        qemu       qcow2.xz
x86_64        vmware     ova
```

**Comment by dustymabe** [comment]
> There are a few things left to do before it's all plumbed through:
>
> * plumb `metal4k` through all the metadata and website
> * adjust coreos-installer to automatically use the `metal4k` image

Some work has happened since then. Is the above still the current status?

**Comment by jlebon** [comment]
Yes, this is done. The download page already lists 4k images but I've submitted a patch to give it a better label and have it listed near the regular raw image: https://pagure.io/fedora-web/websites/pull-request/104

**Comment by bgilbert** [comment]
Let's leave this open until coreos-installer can automatically select 4k images and/or we've documented how to install them.

**Comment by jlebon** [comment]
Oh right, I totally forgot about that part. (I actually did start on that, which is probably why I got mixed up and thought it was done already!)

**Comment by bgilbert** [comment]
coreos-installer feature request: https://github.com/coreos/coreos-installer/issues/201

**Comment by jlebon** [comment]
Let's also add CI coverage to the list of things missing before closing this out completely.

**Comment by jlebon** [comment]
Kola support for testing metal4k added in https://github.com/coreos/coreos-assembler/pull/1312 and part of coreos-assembler CI. coreos-installer support and CI testing added in https://github.com/coreos/coreos-installer/pull/203 and pipeline testing in https://github.com/coreos/fedora-coreos-pipeline/pull/218.

So we just need a new coreos-installer release now. @dustymabe Is the new procedure here to leave this open until that happens or do we close it and tag it with something?

**Comment by dustymabe** [comment]
The fix for this landed upstream. It is now pending a testing stream release.

**Comment by dustymabe** [comment]
we're still waiting on that `coreos-installer` release for this right?

**Comment by jlebon** [comment]
> we're still waiting on that `coreos-installer` release for this right?

Yes, correct.

**Comment by dustymabe** [comment]
The fix for this went into testing stream release `31.20200505.2.0`.  Please try out the new release and report issues.

**Comment by dustymabe** [comment]
The fix for this went into stable stream release `31.20200505.3.0`.

**Comment by majorinche** [comment]
is this fixed, i still encounter this issue, install os from pxe env.
-----
the pxelinux.cfg file is:
DEFAULT pxeboot
TIMEOUT 20
PROMPT 0
LABEL pxeboot
    KERNEL fedora-coreos-31.20200505.3.0-live-kernel-x86_64
    APPEND ip=dhcp rd.neednet=1 initrd=fedora-coreos-31.20200505.3.0-live-initramfs.x86_64.img coreos.inst=yes coreos.inst.ignition_url=http://10.30.0.109/example.ign coreos.inst.image_url=http://10.30.0.109/fedora-coreos-31.20200505.3.0-metal.x86_64.raw.xz coreos.inst.install_dev=/dev/sda coreos.inst.insecure=yes
IPAPPEND 2
----


the report error still be:
-----
coreos-installer install /dev/sda --ignition /tmp/coreos-installer-hmtgYF --firstboot-args rd.neednet=1 ip=dhcp BOOTIF=01-3c-2a-1d-da-44 --image-url http://10.30.0.109/fedora-coreos-31.20200505.3.0-metal.x86_64.raw.xz --insecure

**Error: parsing arguments
Caused by: getting sector size of /dev/sda
Caused by: opening "/dev/sda"
Caused by: No such file or directory(os error2)**
-----

**Comment by dustymabe** [comment]
Assuming `/dev/sda` exists in your machine that error message could clearly use some work.

I see you are using the wrong metal image for a 4k sector drive. Could you try with the 4k metal image? `fedora-coreos-31.20200517.3.0-metal4k.x86_64.raw.xz`

Though
